torch/onnx/_internal/exporter.py

```diff
@@ -0,0 +1,248 @@
+        *,
+        serializer: Optional[ExportOutputSerializer] = None,
+    ) -> None:
+        """Saves the in-memory ONNX model to ``destination`` using specified ``serializer``.
+        If no ``serializer`` is specified, the model will be serialized as Protobuf."""
+
+        if serializer is None:
+            serializer = ProtobufExportOutputSerializer()
+        if isinstance(destination, str):
+            with open(destination, "wb") as f:
+                serializer.serialize(self, f)
+        else:
+            serializer.serialize(self, destination)
+
+
+class Exporter(abc.ABC):
+    @_beartype.beartype
+    def __init__(
+        self,
+        options: ExportOptions,
+        model: torch.nn.Module,
+        model_args: Sequence[Any],
+        model_kwargs: Mapping[str, Any],
+    ):
+        self.options = options
+        self.model = model
+        self.model_args = model_args
+        self.model_kwargs = model_kwargs
+
+    @abc.abstractmethod
+    def run(self) -> ExportOutput:
+        pass
+
+    @property
+    def logger(self) -> logging.Logger:
+        # options.logger will always be resolved to an instance when constructing
+        assert isinstance(self.options.logger, logging.Logger)
+        return self.options.logger
+
+
+@_beartype.beartype
+def _resolve_export_options(options: Optional[ExportOptions]) -> ExportOptions:
+    if options is None:
+        options = ExportOptions()
+
+    T = TypeVar("T")
+
+    def resolve(value: Optional[T], fallback: Callable[[], T]):
+        return fallback() if value is None else value
+
+    return ExportOptions(
+        opset_version=resolve(options.opset_version, lambda: _DEFAULT_OPSET_VERSION),
+        dynamic_shapes=options.dynamic_shapes,
+        logger=resolve(
+            options.logger, lambda: logging.getLogger().getChild("torch.onnx")
+        ),
+    )
+
+
+@_beartype.beartype
+def dynamo_export(
+    model: torch.nn.Module,
+    /,
+    *model_args,
+    export_options: Optional[ExportOptions] = None,
+    **model_kwargs,
+) -> ExportOutput:
+    """Export a torch.nn.Module to an ONNX graph.
+
+    Args:
+        model: The PyTorch model to be exported to ONNX.
+        model_args: Positional inputs to ``model``.
+        model_kwargs: Keyword inputs to ``model``.
+        export_options: Options to influence the export to ONNX.
+
+    Returns:
+        An in-memory representation of the exported ONNX model.
+
+    Example:
+    ::
+
+        import torch.onnx
+        torch.onnx.dynamo_export(
+            my_nn_module,
+            torch.randn(2, 2, 2), # positional input 1
+            torch.randn(2, 2, 2), # positional input 2
+            my_nn_module_attribute="hello", # keword input
+            export_options=ExportOptions(
+                opset_version=17,
+            )
+        ).save("my_model.onnx")
+    """
+
+    resolved_export_options = _resolve_export_options(export_options)
+
+    def assert_dependencies():
```

Comment:
Curious what are the messages we display to users when dependencies are missing?

Reply:
This depends on Phase 2: Diagnostics. @BowenBao will own that. Messages will be granular and actionable.